repository/httpRemoteRepository.go

```diff
@@ -0,0 +1,296 @@
+	cacheDirectory := filepath.Join(r.endpointCacheDir, name.String())
+	return r.receiveDoguFromRemoteOrCache(requestUrl, cacheDirectory)
+}
+
+// Get returns a version specific detail about the dogu.
+func (r *httpRemote) Get(_ context.Context, doguVersion dogu.QualifiedVersion) (*core.Dogu, error) {
+	err := doguVersion.Name.Validate()
+	if err != nil {
+		return nil, fmt.Errorf("qualified dogu name is not valid (name: %s): %w", doguVersion.Name.String(), err)
+	}
+	requestUrl := r.urlSchema.GetVersion(doguVersion.Name.String(), doguVersion.Version.Raw)
+	cacheDirectory := filepath.Join(r.endpointCacheDir, doguVersion.Name.String(), doguVersion.Version.Raw)
+	return r.receiveDoguFromRemoteOrCache(requestUrl, cacheDirectory)
+}
+
+func (r *httpRemote) receiveDoguFromRemoteOrCache(requestUrl string, dirname string) (*core.Dogu, error) {
+	var remoteDogu, err = r.readCachedDogu(dirname)
+	if err != nil {
+		err = retry.OnError(maxTries, isRetryError, func() error {
+			remoteDogu, err = r.request(requestUrl, true)
+			return err
+		})
+
+		if err != nil {
+			return nil, err
+		}
+
+		err = r.writeDoguToCache(remoteDogu, dirname)
+		if err != nil {
+			return &core.Dogu{}, fmt.Errorf("failed to write dogu to cache: %w", err)
+		}
+	}
+
+	return remoteDogu, nil
+}
+
+func (r *httpRemote) readCachedDogu(dirname string) (*core.Dogu, error) {
+	if r.useCache {
+		cacheFile := filepath.Join(dirname, "content.json")
+		doguFromFile, _, err := core.ReadDoguFromFile(cacheFile)
+		if err != nil {
+			return nil, commonerrors.NewGenericError(fmt.Errorf("failed to read from cache %s: %w", cacheFile, err))
+		}
+		if doguFromFile == nil {
+			return nil, commonerrors.NewNotFoundError(fmt.Errorf("dogu descriptor not found"))
+		}
+		return doguFromFile, nil
+	}
+	return nil, commonerrors.NewGenericError(fmt.Errorf("useCache is not activated"))
+}
+
+func (r *httpRemote) writeDoguToCache(doguToWrite *core.Dogu, dirname string) error {
+	err := os.MkdirAll(dirname, os.ModePerm)
+	if nil != err {
+		return fmt.Errorf("failed to create cache directory %s: %w", dirname, err)
+	}
+
+	cacheFile := filepath.Join(dirname, "content.json")
+	err = core.WriteDoguToFile(cacheFile, doguToWrite)
+
+	if nil != err {
+		removeErr := os.Remove(cacheFile)
+		if removeErr != nil {
+			core.GetLogger().Warningf("failed to remove cache file %s", cacheFile)
+		}
+		return fmt.Errorf("failed to write cache %s: %w", cacheFile, err)
+	}
+
+	return nil
+}
+
+func (r *httpRemote) request(requestURL string, useCredentials bool) (*core.Dogu, error) {
+	core.GetLogger().Debugf("fetch json from remote %s", requestURL)
+
+	request, err := http.NewRequest("GET", requestURL, nil)
+	if err != nil {
+		return nil, commonerrors.NewGenericError(fmt.Errorf("failed to prepare request: %w", err))
+	}
+
+	if useCredentials && r.credentials != nil {
+		request.SetBasicAuth(r.credentials.Username, r.credentials.Password)
+	}
+
+	resp, err := r.client.Do(request)
```

Comment:
🚫 **[checkstyle]** <bodyclose> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>response body must be closed
<!-- __reviewdog__:ChA4MTcyMzBiOWJlNWI2M2MzEgpjaGVja3N0eWxl -->